File: Python/PyLife/wanderer.py
from random import randint

from colors import ORANGE
from draw_handler import DrawData
from grid_point import GridPoint
from moveable import Moveable
from polygon import DIAMOND, TRIANGLE_N, TRIANGLE_E, TRIANGLE_S, TRIANGLE_W
from velocity import DIR_UP, DIR_RIGHT, DIR_DOWN, DIR_LEFT, DIR_NONE


class Wanderer(Moveable):
    def __init__(self, pos: GridPoint = None, wait_prob=1):
        super().__init__(pos, None)
        self._max_rand = 3
        self._direction = DIR_NONE
        self._draw_data = DrawData(DIAMOND, ORANGE, False)

        if wait_prob > 1:
            self._max_rand += wait_prob

    def draw(self) -> DrawData:
        if self._direction == DIR_UP:
            poly = TRIANGLE_N
        elif self._direction == DIR_RIGHT:
            poly = TRIANGLE_E
        elif self._direction == DIR_DOWN:
            poly = TRIANGLE_S
        elif self._direction == DIR_LEFT:
            poly = TRIANGLE_W
        else:
            poly = DIAMOND

        self._draw_data.polygon = poly

        return self._draw_data

    def move(self):
        self._direction = randint(0, self._max_rand)

        if self._direction == DIR_UP:
            dx = 0
            dy = -1
        elif self._direction == DIR_RIGHT:
            dx = +1
            dy = 0
        elif self._direction == DIR_DOWN:
            dx = 0
            dy = +1
        elif self._direction == DIR_LEFT:
            dx = -1
            dy = 0
        else:
            dx = 0
            dy = 0

        self._velocity.dx = dx
        self._velocity.dy = dy
        super().move()

File: JavaScript/JSLife/wanderer.js
const _W_DIRS = {
  NONE: 0,
  N: 1,
  E: 2,
  S: 3,
  W: 4,
};
const _W_NUM_DIRS = Object.keys(_W_DIRS).length;
const _W_DRAW_DATA = new DrawData(DIAMOND, '#ff0000', false);

class Wanderer extends Moveable {
  constructor(x = 0, y = 0, speed = 1, sleepProbability = 0) {
    super(x, y, null, null);
    this._speed = speed;
    this._drawData = _W_DRAW_DATA.clone();
    this._computeVelocity(this._generateDirection());
    this._sleepProb = sleepProbability;
  }
  
  _computeVelocity(direction) {
    switch (direction) {
      case _W_DIRS.N:
        this._dx = 0;
        this._dy = -this._speed;
        this._drawData.polygon = TRIANGLE_N;
        break;
      
      case _W_DIRS.E:
        this._dx = +this._speed;
        this._dy = 0;
        this._drawData.polygon = TRIANGLE_E;
        break;
      
      case _W_DIRS.S:
        this._dx = 0;
        this._dy = +this._speed;
        this._drawData.polygon = TRIANGLE_S;
        break;
      
      case _W_DIRS.W:
        this._dx = -this._speed;
        this._dy = 0;
        this._drawData.polygon = TRIANGLE_W;
        break;
      
      default:
        this._dx = 0;
        this._dy = 0;
        this._drawData.polygon = DIAMOND;
    }
  }
  
  _generateDirection() {
    return randomInt(-5, _W_NUM_DIRS - 1);
  }
  
  move() {
    super.move();
    let dir = this._generateDirection();
    this._computeVelocity(dir);
  }
  
  draw() {
    return this._drawData;
  }
}
